flow/aggregates.py

```diff
@@ -0,0 +1,481 @@
+        common state point parameter 'sp' whose value, when not found, is replaced by a
+        default value of -1.
+
+        .. code-block:: python
+
+            @aggregator.groupby(key='sp', default=-1)
+            @FlowProject.operation
+            def foo(*jobs):
+                print(len(jobs))
+
+        :param key:
+            Parameter which specifies how to group jobs based on a single
+            or a list of state point parameters or a custom callable.
+        :type key:
+            str, Iterable, or callable
+        :param default:
+            Default values used for grouping if invalid key is passed.
+        :type default:
+            str, Iterable, or callable
+        :param sort_by:
+            Before aggregating, sort the jobs by a given statepoint parameter.
+            The default behavior is no sorting.
+        :type sort_by:
+            str or NoneType
+        :param reverse_order:
+            States if the jobs are to be sorted in reverse order.
+            The default value is False.
+        :type reverse_order:
+            bool
+        :param select:
+            Condition for filtering individual jobs. This is passed as the
+            callable argument to `filter`.
+            The default behavior is no filtering.
+        :type select:
+            callable or NoneType
+        """
+        if isinstance(key, str):
+            if default is None:
+                def keyfunction(job):
+                    return job.sp[key]
+            else:
+                def keyfunction(job):
+                    return job.sp.get(key, default)
+        elif isinstance(key, Iterable):
+            keys = list(key)
+            if default is None:
+                def keyfunction(job):
+                    return [job.sp[key] for key in keys]
+            else:
+                if isinstance(default, Iterable):
+                    if len(default) != len(keys):
+                        raise ValueError("Expected length of default argument is "
+                                         f"{len(keys)}, got {len(default)}.")
+                else:
+                    raise TypeError("Invalid default argument. Expected Iterable, "
+                                    f"got {type(default)}")
+
+                def keyfunction(job):
+                    return [job.sp.get(key, default_value)
+                            for key, default_value in zip(keys, default)]
+        elif callable(key):
+            keyfunction = key
+        else:
+            raise TypeError("Invalid key argument. Expected either str, Iterable "
+                            f"or a callable, got {type(key)}")
+
+        def aggregator_function(jobs):
+            for key, group in groupby(sorted(jobs, key=keyfunction), key=keyfunction):
+                yield group
+
+        return cls(aggregator_function, sort_by, reverse_order, select,
+                   key=key, default=default)
+
+    def __eq__(self, other):
```

Comment:
If we make the change to `_MakeAggregates` I suggest, then we should move the `__eq__` and `__hash__` methods to it.

Reply:
I'm not sure I see the use case of moving them to `_MakeAggregates` can you please explain?

Reply:
We'll be eventually storing the instances of `_AggregateStore` and not `_MakeAggregates`. Hence, I think it is not useful to move `__eq__` and `__hash__` method to `_MakeAggregate`

Reply:
Nevermind, we do need this at first to be able to discern if an user creates multiple of the same `aggregator` objects... We will likely need a way to hash and equate `_AggregateStores` as well though.

Reply:
@b-butler 
In this implementation, we do have an `__eq__` and `__hash__` methods for both `_DefaultAggregateStore` and `_AggregatesStore`

Reply:
Yep.